[export] basic auto dynamic shapes

pianpwk opened the pull request:
Starter version of automatic dynamic shapes for export.

Creates enums `DIM.AUTO`, `DIM.STATIC`, allowing user to specify `AUTO` for dims in dynamic_shapes specs, meaning that corresponding dims are treated as dynamic, and relevant guards will do what's necessary (e.g. refine ValueRanges, set replacements based on equality, or even set static) without raising ConstraintViolationErrors. Basically allows the user to say, "a bunch of these dims can be dynamic, let export do model analysis and return the program with maximum possible dynamism, without complaining".

The usage for specifying `dynamic_shapes` is now:
```
AUTO -> dynamic by default, return whatever produce_guards() says, even if it's static
None/int/STATIC -> static
Dim/DerivedDim -> same as before - will complain if the min/max range is invalid, or if dims related to this are unspecified.
```

Caveat 1: specifying `AUTO` for a dim won't guarantee it'll be dynamic:

- specifying `AUTO` for a dim will return the maximum possible dynamism given your program and other specified constraints, but this can still mean you'll get a static program. For example, with the program below, x is specified dynamic, but it's equal to y, which is specified static, and with how we currently do things we won't promote y to dynamic, but will demote(?) x to static. So this can be surprising if you don't fully know your model, and/or missed one of your other inputs when specifying auto-dynamic shapes.
```
class Foo(torch.nn.Module):
    def forward(self, x, y):
        return x + y
inputs = (torch.randn(6), torch.randn(6))
export(Foo(), inputs, dynamic_shapes={"x": (DIM.AUTO,), "y": None})
```

Caveat 2: specifying `AUTO` and Dims in the same spec is still problematic:

- The way Dims/DerivedDims are currently handled is very strict. A Dim represents a symbol, and we require a user to specify the symbol for all dims governed by the symbol - that's why we've seen errors in the past like `The values of x must always be related to y by ...`, asking the user to specify the exact relation as in the program. We also require the specified min/max range to be a subset of the valid range from model analysis. All this doesn't compose well with specifying `AUTO` just yet - for example in the program below, ideal behavior could be to return a dynamic program, where `dx = x.size(0) = y.size(0)` has range (3,6). Unfortunately this crashes, and correct behavior is to specify `dx` for both inputs. So currently we raise a UserError and crash if both Dims + `AUTO` are present in the spec.
```
class Foo(torch.nn.Module):
    def forward(self, x, y):
        return x + y
inputs = (torch.randn(6), torch.randn(6))
export(Foo(), inputs, dynamic_shapes={"x": (DIM.AUTO,), "y": {0: Dim("dx", min=3, max=6)}})  # this doesn't work, because x & y and related
```

Implementation details:

This is done by setting `assume_static_by_default=False`, and doing a transform on the `dynamic_shapes` spec to preserve semantics. `assume_static_by_default=False` will treat unspecified dims or Nones as dynamic. This is the opposite of what `export.export()` currently does - unspecified Dims/Nones are treated as static. Historically this static-by-default behavior, where the user deals with fewer guards, has been desirable, and we would like to respect that in this implementation. So this internal spec transformation is added, `_transform_shapes_for_default_dynamic()`, does the spec conversion necessary to be compatbile with dynamic by default. Specifically, AUTOs are converted into Nones, and Nones/unspecified dims are filled in with explicitly static constraints. 

For example, this would look like, for a 3-d tensor: `{0: DIM.AUTO, 1: None, 2: Dim("dx")} -> {0: None, 1: 32, 2: Dim("dx")}`

This does seem overly complicated, but it's done to preserve dynamic shapes semantics for `torch._dynamo.export()`, which already uses `assume_static_by_default=False`, and follows the same process for generating shape constraints , via `_process_dynamic_shapes`. There the semantics are:
```
None/unspecified: dynamic by default
Dim/DerivedDim: also a strict assertion
```

If we don't care about BC for `_dynamo.export(dynamic_shapes)`, then we can just modify semantics for `_process_dynamic_shapes()` and change all the relevant tests in `test/dynamo/test_export.py`.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

pianpwk commented on test/export/test_export.py:
```diff
@@ -4269,8 +4296,6 @@ def forward(self, xs, y):
     @testing.expectedFailureSerDer  # We don't preserve metadata on graph module
-    @testing.expectedFailureNonStrict
-    @testing.expectedFailureTrainingIRToRunDecompNonStrict
     def test_retrace_graph_level_meta_preservation(self):
```
It seems like now we just consistently store static constraints for unspecified dims, like `{"t_id": ..., "dim": ..., "min": 4, "max": 4} so this doesn't break anymore. Any thoughts on this?

pianpwk commented on test/export/test_export.py:
```diff
@@ -4396,6 +4421,7 @@ def forward(self):
         ep = export(m, ())
         self.assertEqual(ep.graph_signature.lifted_tensor_constants, ["x"])
 
+    @unittest.expectedFailure
```
This PR seems to expose some already existing problem between explicitly static constraints + dataclasses (or equality constraints)? Marking this as a failure to open the rest of the PR for review, but will work on this and resolve it before this PR is merged in.

avikchaudhuri replied:
OK, as long as you get to it. I'd guess info on unused inputs could easily be perturbed.

pianpwk replied:
about to add a fix in the next commit

pianpwk commented on test/export/test_export.py:
```diff
@@ -6787,6 +6813,178 @@ def forward(self, x):
             if node.op == "call_function":
                 self.assertTrue(False)
 
+    def test_automatic_dynamic_shapes_simple_equality(self):
```
these 3 tests explain a lot of the behavior for this API

pianpwk commented on torch/_dynamo/eval_frame.py:
```diff
@@ -1307,7 +1303,6 @@ def export(
     def inner(*args, **kwargs):
         combined_args = _combine_args(_f, args, kwargs)
-        _check_dynamic_shapes(combined_args, dynamic_shapes)
         constraints = _process_dynamic_shapes(combined_args, dynamic_shapes)
```
Move this to _export_to_torch_ir

avikchaudhuri replied:
Meaning what, the errors happen earlier or later?

pianpwk replied:
earlier, export_to_torch_ir() will check the original spec, and do the transform

pianpwk commented on torch/_export/non_strict_utils.py:
```diff
@@ -140,11 +141,7 @@ def make_fake_inputs(
-            len(constraints) == 0
-        ), "Found constraints when tracing with a toplevel tracing context."
+        # create another fake mode.
         fake_mode = context.fake_mode
```
cc @angelayi

avikchaudhuri replied:
No idea what's up with the previous comment here. 

If this assertion passed before, then `dynamic_shapes` should have been empty, no? So that means that we will always now create a full `dynamic_shapes` for inputs, and thus constraints may not be empty: in particular, everything will be explicitly static. That's why we need to remove this assertion, yeah?

But what's worrying me about the comment is that, what would break if the constraints are no longer empty, is there some kind of "shadowing" done by "toplevel tracing context"?

pianpwk replied:
Yeahh I don't know what's going on here, but @angelayi said it should be fine to remove this.

It was breaking for this test case, where we torch compile with export as backend?

pianpwk commented on torch/_export/non_strict_utils.py:
```diff
@@ -339,21 +340,25 @@ def make_constraints(
                 # and store it indexed by the symbolic expression corresponding to it.
                 # NOTE(avik): Use node._expr instead of node.expr for the lookup here because
                 # we want the symbol, not its replacement, which could be an expression. Maybe
                 # there's a better way to do this, e.g., by (re)computing value ranges for expressions?
```
just swapping if/else, so the condition is easier to understand

pianpwk commented on torch/export/_trace.py:
```diff
@@ -1819,7 +1828,9 @@ def _produce_guards_callback(gm):
+    with fake_mode, _DataDependentErrorHandlerNonStrict(), torch._dynamo.config.patch(
+        assume_static_by_default=False
+    ):
         with _fakify_script_objects(mod, fake_args, fake_kwargs, fake_mode) as (
```
Wrapped around _export_to_aten_ir

pianpwk commented on torch/export/dynamic_shapes.py:
```diff
@@ -904,6 +909,149 @@ def check_shape(path, t, dynamic_shape):
+            )
+
+
+def _transform_shapes_for_default_dynamic(
```
This entire function is unnecessary if we could ignore BC and completely change `_dynamo.export()` behavior, but this seems safer to start with.

avikchaudhuri replied:
I think it is better to make the full spec explicit early in the export pathway anyway, rather than carry the meanings of defaults and interpret them everywhere. Makes it easier to adapt to automated use cases where we can just generate complete specs from side info (in addition to continuing to support user-defined specs, which have some pressure to be minimal).

avikchaudhuri reviewed (commented):
Haven't looked at the code yet, but just want to record some high-level comments so far:
* Great summary!
* I don't like `True`, and I hate `None`. For `True`, instead I want something like `torch.export.Dim.AUTO` (ideally an enum we can grow). For `None`, you can do this in a follow up but I'd really like to have a distinction between the two roles played by it, something like `torch.export.Dim.STATIC` in the future, and the fact that you can not bother with specifying a subtree (which can remain `None`), and it would align with not specifying dynamic shapes at all (implicitly `None`).
* For interactions of `True` with static, say in `x + y`, I think we should move to a design where the exported program records the dynamic shape spec explicitly and any specialization would just show up there for those who care. Warnings are a bonus.
* For interactions of `True` with `Dim`, likewise we should aspire to constrain the `True` to match `Dim`. This is kind of like how the Python type `Any` would merge with a more informative type. Then the interaction with static would match the semantics in principle: `True` starts of with minimal assumptions and gets more constrained over time.

SherlockNoMad commented on test/export/test_export.py:
```diff
@@ -6787,6 +6813,178 @@ def forward(self, x):
+        )
+
+    def test_automatic_dynamic_shapes_constant_relation(self):
+        # case 2: related by constant: s0 + 4 = s1
```
nice!

avikchaudhuri commented on test/export/test_export.py:
```diff
@@ -6787,6 +6813,178 @@ def forward(self, x):
+        self._check_dynamic_shapes_specs_and_shapes(
+            SimpleEquality(),
+            inputs,
+            specs=[
```
lol now I'm convinced `True` should be renamed. ;)

avikchaudhuri commented on torch/_export/non_strict_utils.py:
```diff
@@ -85,15 +86,14 @@ def fakify(
     assert mode.shape_env is not None
     if t_id in t_constraints:
         for i, constraint in t_constraints[t_id].items():
             symbolic_context.constraint_sizes[i] = constraint.constraint_range
```
So what about static dims? They will have a singleton constraint range?

pianpwk replied:
Yepp, but we will convert them into static constraints at some point, just not here (I think before we run into any guards).

pianpwk replied:
Yeah that's right, there's [some code](https://github.com/pytorch/pytorch/blob/416a2edd7c494bc6d739c27ddd538cf2d1a94d74/torch/fx/experimental/symbolic_shapes.py#L2629-L2639) that'll convert these into static later on though, so they won't have symbols.

But ok, looks like this will still create a sympy.Integer instead of an int, so this is probably responsible for all the is_number checks.

avikchaudhuri commented on torch/_export/non_strict_utils.py:
```diff
@@ -315,6 +312,10 @@ def make_constraints(
     }
+    if dynamic_shapes is True:
+        dynamic_shapes = _tree_map_with_path(
+            lambda path, t, shape: None, combined_args, combined_args
```
So `None` for an entire tensor, right? Not `None` for individual dims. This will somehow cause everything to be marked auto-dynamic? Can you link to the code that does that?

avikchaudhuri replied:
Add a comment too, please.

pianpwk replied:
Oh, the code that does that is just process_dynamic_shapes doing nothing for Nones, with assume_static_by_default=False

tugsbayasgalan replied:
Sorry I think I asked this before, but I forgot what the answer was. What happens when you specify True and also specify Dim constraints for the shape? Do we error or do we intersect them?

tugsbayasgalan replied:
Ahh we error, nvm

avikchaudhuri commented on torch/_export/non_strict_utils.py:
```diff
@@ -339,21 +340,25 @@ def make_constraints(
         shape_spec = flat_dynamic_shapes[input_index - num_lifted_inputs]
         for i, d in enumerate(node.meta["val"].shape):
-            if isinstance(d, torch.SymInt):
+            if isinstance(d, torch.SymInt) and not d.node.expr.is_number:
```
What happens when `d.node.expr.is_number`? Maybe a positive test would be better here, are you expecting it to be a symbol or are you worried it could be an expression containing symbols?

avikchaudhuri replied:
Also looking at my comment in the code, I'm wondering if `_expr` could be a symbol but `expr` is a number, maybe because a replacement exists? Not sure, babbling here, maybe you observed something that can help refresh my memory?

pianpwk replied:
It seems like auto dynamic mode, when things specialize, will create a torch.symint instead of a concrete int

avikchaudhuri commented on torch/export/_trace.py:
```diff
@@ -551,8 +560,8 @@ def _export_to_torch_ir(
-                    dynamic_shapes=dynamic_shapes,  # type: ignore[arg-type]
-                    assume_static_by_default=True,
+                    dynamic_shapes=_dynamic_shapes,  # type: ignore[arg-type]
+                    assume_static_by_default=False,
```
I think you can just drop this, it is `False` by default.

avikchaudhuri commented on torch/export/dynamic_shapes.py:
```diff
@@ -805,8 +805,13 @@ def _check_dynamic_shapes(
 
-    if dynamic_shapes is None or len(dynamic_shapes) == 0:
+    if (
+        dynamic_shapes is True
```
You sure you need the `True` here? The `None` value plays a couple of roles here, one for subtrees (including tensors as a whole) and one for dimensions. I believe you want `True` to apply only to dimensions, like `Dim`.

pianpwk replied:
Yeah, this is in case of `dynamic_shapes=True` - we'll let everything be dynamic

avikchaudhuri commented on torch/export/dynamic_shapes.py:
```diff
@@ -832,32 +837,32 @@ def check_symbols(path, tensor, shape):
                         case_name="dynamic_shapes_validation",
                     )
-        elif shape is not None:
+        elif shape not in [None, True]:
```
Likewise, I don't think it should ever be `True` here, unless you are thinking of a different role for `True` here that applies to subtrees.

pianpwk replied:
Yeah I mentioned above, this allows True to apply to the entire structure, a tensor, or a Dim

avikchaudhuri commented on torch/export/dynamic_shapes.py:
```diff
@@ -904,6 +909,149 @@ def check_shape(path, t, dynamic_shape):
+        Non-builtin types for pytree (e.g. custom dataclasses) creates some difficulty,
+        in which case the correct format is a list containing specs for each child attribute.
+        """
+        if (node_type := _get_node_type(tree)) not in SUPPORTED_NODES:  # is_leaf
```
There's some similar logic in `_tree_map_with_path`, have you cross-checked? In particular, for registered pytree nodes we match whatever `flatten` gives. Since you mentioned a failing test that involves dataclasses, this could be an area where there's maybe something extra or missing wrapping?

pianpwk replied:
Yeah, I guess this is basically _tree_map_with_path + we want to unflatten as a list when the type isn't builtin - for some reason that's the correct spec for dynamic_shapes

avikchaudhuri commented on torch/export/dynamic_shapes.py:
```diff
@@ -904,6 +909,149 @@ def check_shape(path, t, dynamic_shape):
+        combined_args = type(dynamic_shapes)(combined_args.values())  # type: ignore[assignment, misc]
+
+    def transform_shapes(path, tensor, shape):
+        def _marked_dynamic(tensor, i):
```
Oh boy, we have to handle these as well in export? I'd really prefer not to. If you have to, can you describe the semantics here (also add comment)?

pianpwk replied:
Yeah, it's used in some AOTInductor test cases, you could leave `dynamic_shapes` as None, but call `mark_dynamic(tensor, index=i)`

tugsbayasgalan replied:
Hmm sorry might be a dumb question: couldn't we rewrite AOTInductor cases to use the proper dynamic shapes API? I am afraid we are making this logic too complicated hahaa

avikchaudhuri replied:
I agree. I don't think we want to be in the business of defining what `mark_dynamic` does in export: it looks like it has the same effect as `True`? So in principle it should be possible to update those tests. If that's too much scope for this PR, we should definitely track a follow up.

pianpwk replied:
There was a small moment I was afraid all the changes in this PR would be unnecessary, and we only needed to switch to calling `mark_dynamic` under the hood to implement auto dynamic shapes, but seems like currently it is different from `True`, and slightly different from creating a plain `Dim()`:

- Unlike True, it doesn't automatically specialize or constrain itself, so you'll get a constraint violation
- Unlike Dim, if you mark multiple dims dynamic and they turn out to be equal, they'll get the symbol without complaining

There seems to be a lot of tests that rely on this (AOTInductor, torchbench, probably some internal), so I'll try to maintain support as is for this PR, and then we can decide to deprecate after?

avikchaudhuri replied:
Sounds good.

avikchaudhuri replied:
Could you leave a TODO in the code and file a task?

avikchaudhuri commented on torch/export/dynamic_shapes.py:
```diff
@@ -904,6 +909,149 @@ def check_shape(path, t, dynamic_shape):
+        def _marked_dynamic(tensor, i):
+            return i in getattr(tensor, "_dynamo_dynamic_indices", set())
+
+        if isinstance(shape, dict):
```
We should really have a separate util class that does things with shapes of different kinds, like tuple or dict or None, so that we can avoid some boilerplate.

avikchaudhuri replied:
Could you file a task on this?

avikchaudhuri commented on torch/export/dynamic_shapes.py:
```diff
@@ -914,7 +1062,7 @@ def _process_dynamic_shapes(
     from torch._dynamo.exc import UserError, UserErrorType
 
-    if dynamic_shapes is None or len(dynamic_shapes) == 0:
+    if dynamic_shapes is True or dynamic_shapes is None or len(dynamic_shapes) == 0:
```
Again, maybe not necessary?

pianpwk commented on torch/_dynamo/eval_frame.py:
```diff
@@ -1623,14 +1625,6 @@ def graph_with_interpreter(*args):
                 result_traced,  # type: ignore[possibly-undefined]
                 flat_args_dynamic_dims,
             )
-        # Store constraints and inputs as metadata for user passes, e.g. turn constraints to runtime check
```
killing this metadata

pianpwk commented on torch/_dynamo/eval_frame.py:
```diff
@@ -1608,7 +1603,14 @@ def graph_with_interpreter(*args):
 
         if same_signature:
             flat_args_dynamic_dims = [
-                {c.dim for c in (constraints or ()) if c.w_tensor() is x}
```
needed to keep dims on unused shapes as static

avikchaudhuri replied:
you'll probably hit a merge conflict here, I changed this to use `c.t_id` instead

pianpwk commented on test/dynamo/test_export.py:
```diff
@@ -2756,36 +2756,6 @@ def forward(self, x, y):
             ["torch.Size([s0, s1, s1])", "torch.Size([s0, s1, s1])"],
         )
 
-    @config.patch(
```
sorry, we're killing this metadata

pianpwk commented on test/export/test_export.py:
```diff
@@ -4268,47 +4295,6 @@ def forward(self, xs, y):
                 len([node for node in gm.graph.nodes if node.op == "placeholder"]), 2
             )
 
-    @testing.expectedFailureSerDer  # We don't preserve metadata on graph module
```
killing metadata here

pianpwk commented on torch/_dynamo/eval_frame.py:
```diff
@@ -1568,7 +1564,14 @@ def graph_with_interpreter(*args):
+                        c.t_id == id(x)
+                        and c.constraint_range.vr.lower != c.constraint_range.vr.upper
+                    )
+                }
```
needed so unused tensor shapes remain static

pianpwk commented on torch/export/dynamic_shapes.py:
```diff
@@ -993,7 +1140,7 @@ def root_value():
     def update_symbols(path, tensor, shape):
         def _create_static_dim(tensor, i, value):
-            return _StaticDim(str(value), (int,), {"value": value})
+            return _StaticDim(f"{value}_{id(tensor)}_{i}", (int,), {"value": value})
```
having the same name for all shapes of a certain value was creating too many unnecessary equality constraints, I'm switching to avoid it

pianpwk replied:
Noo, there's some code to clean up these fake dynamic constraints before produce_guards(), it just converts them into static constraints

pianpwk replied:
Ok turns out it does, gonna revert this change...

avikchaudhuri commented on test/export/test_export.py:
```diff
@@ -275,6 +275,21 @@ def _test_export_same_as_eager(self, f, args, kwargs=None):
+                ep.module()(*test_inputs)
+            for shapes in failing_shapes:
+                test_inputs = (torch.randn(*shape) for shape in shapes)
+                with self.assertRaisesRegex(RuntimeError, ""):
```
`assertRaises` or be more specific

tugsbayasgalan commented on torch/_export/non_strict_utils.py:
```diff
@@ -125,11 +125,12 @@ def make_fake_inputs(
-    combined_args = _combine_args(
-        nn_module, args, kwargs, _is_torch_jit_trace=_is_torch_jit_trace
-    )
+    combined_args = _combine_args(nn_module, args, kwargs)
```
Why do we not have _is_torch_jit_trace flag anymore? This was needed when we retrace torch.jit.trace-d modules to export.

pianpwk replied:
Oh, for this callsite it's not needed, we just need the dictionary binding

avikchaudhuri reviewed (commented):
So what's the resolution on `True`? It's uninformative. Also it seems you want it to apply to entire tensors, is that right?

tugsbayasgalan commented on torch/export/_trace.py:
```diff
@@ -1819,7 +1827,9 @@ def _produce_guards_callback(gm):
         )
 
-    with fake_mode, _DataDependentErrorHandlerNonStrict():
+    with fake_mode, _DataDependentErrorHandlerNonStrict(), torch._dynamo.config.patch(
```
Don't we only set this to False if there is some shape marked with "True"?

pianpwk replied:
Ah no, this PR redoes the whole spec, creating static constraints for everything not marked True or Dim

avikchaudhuri commented on test/export/test_export.py:
```diff
@@ -4268,47 +4295,6 @@ def forward(self, xs, y):
                 len([node for node in gm.graph.nodes if node.op == "placeholder"]), 2
             )
 
-    @testing.expectedFailureSerDer  # We don't preserve metadata on graph module
```
`input_shape_constraints` is internal-only, no users. I'm planning to change it as well.

avikchaudhuri commented:
> So what's the resolution on `True`? It's uninformative. Also it seems you want it to apply to entire tensors, is that right?

Discussed offline with @pianpwk that (1) `True` should be renamed to `AUTO` (2) it should only apply on dims, not tensors.

pianpwk commented:
For some reason I can't reply to some of @tugsbayasgalan's comments lol:

For the AOTInductor issue: I guess so, I'm just afraid there's internal users relying on `mark_dynamic()` that'll break internal cases. Maybe we can push for that after this.

pianpwk commented on test/export/test_export.py:
```diff
@@ -2446,6 +2490,36 @@ def g(x):
         ):
             constraints = [dynamic_dim(inp_for_g, 0)]
 
+    def test_mark_and_auto_dynamic(self):
```
added test with mark_dynamic. Semantics haven't changed, this is just a test case where AUTO & mark_dynamic result in the same program

pianpwk commented on torch/export/dynamic_shapes.py:
```diff
@@ -533,7 +544,11 @@ def _process_equalities(
+    if not sources:  # empty sources due to unused shapes
+        return
+
+    source, *other_sources = sources
```
this was breaking a test with unused shapes

pianpwk commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -3847,7 +3847,10 @@ def track_symint(source, val, constraint=None):
+                    if (
+                        constraint is not None
+                        and not isinstance(constraint, RelaxedUnspecConstraint)
+                    ):
```
came up with `mark_dynamic()`

avikchaudhuri commented on test/export/test_export.py:
```diff
@@ -2446,6 +2490,36 @@ def g(x):
+        ep_auto = torch.export.export(
+            Foo(), inputs, dynamic_shapes={"x": (AUTO, None), "y": (AUTO, None)}
+        )
+        torch._dynamo.mark_dynamic(inputs[0], 0)
```
Can you say whether `mark_dynamic` only works on dims declared `AUTO` or does it work for anything? Any assertions you want to make here?

pianpwk replied:
Ah it'll work on anything

avikchaudhuri commented on torch/export/dynamic_shapes.py:
```diff
@@ -29,13 +30,23 @@
 ]
 
 
+class DIM(Enum):
```
We have `Dim` already, I was hoping we could hang these constants off that name. I can try in a follow up.

So something like 
```python
import Dim from torch.export
my_tensor_shape = (Dim("batch"), 5, Dim.AUTO)
```

avikchaudhuri commented on torch/export/dynamic_shapes.py:
```diff
@@ -791,8 +806,10 @@ def _check_dynamic_shapes(
+    from torch._export.non_strict_utils import _flatten_dynamic_shapes
 
-    if dynamic_shapes is None or len(dynamic_shapes) == 0:
+    if dynamic_shapes is True or dynamic_shapes is None or len(dynamic_shapes) == 0:
```
`True` not supported any more right?

avikchaudhuri commented on torch/_export/non_strict_utils.py:
```diff
@@ -315,6 +313,11 @@ def make_constraints(
     range_constraints = {
         symbol: inline_constraints[symbol] for symbol in inline_constraints
     }
+    if dynamic_shapes is True:
```
Remove support for `True`?

avikchaudhuri commented on torch/export/dynamic_shapes.py:
```diff
@@ -818,32 +835,32 @@ def check_symbols(path, tensor, shape):
                 if isinstance(dim, _Dim):
                     check_same_bounds(dim)
-                elif not isinstance(dim, int) and dim is not None:
+                elif not (isinstance(dim, (int, DIM)) or dim is None):
```
Or `STATIC`, I think. Can you file a task to convert `None` to `STATIC` in Dim positions?

avikchaudhuri commented on torch/export/dynamic_shapes.py:
```diff
@@ -818,32 +835,32 @@ def check_symbols(path, tensor, shape):
                         case_name="dynamic_shapes_validation",
                     )
-        elif shape is not None:
+        elif shape not in [None, True]:
```
Looks like you're leaving `True` in. We need to make a call here. If you insist then we should document it in the public API docstring. Otherwise remove.

As I argued elsewhere, `True` doesn't really mean that you get a dynamic shape for sure, so what exactly is true with `True` is confusing.

pianpwk replied:
Ah sorry, this was a mistake

avikchaudhuri reviewed (approved):
Some remaining nits but looks good.

pianpwk committed:
[export] basic auto dynamic shapes (#133620)

Summary:
Starter version of automatic dynamic shapes for export.

Creates enums `DIM.AUTO`, `DIM.STATIC`, allowing user to specify `AUTO` for dims in dynamic_shapes specs, meaning that corresponding dims are treated as dynamic, and relevant guards will do what's necessary (e.g. refine ValueRanges, set replacements based on equality, or even set static) without raising ConstraintViolationErrors. Basically allows the user to say, "a bunch of these dims can be dynamic, let export do model analysis and return the program with maximum possible dynamism, without complaining".

The usage for specifying `dynamic_shapes` is now:
```
AUTO -> dynamic by default, return whatever produce_guards() says, even if it's static
None/int/STATIC -> static
Dim/DerivedDim -> same as before - will complain if the min/max range is invalid, or if dims related to this are unspecified.
```

Caveat 1: specifying `AUTO` for a dim won't guarantee it'll be dynamic:

- specifying `AUTO` for a dim will return the maximum possible dynamism given your program and other specified constraints, but this can still mean you'll get a static program. For example, with the program below, x is specified dynamic, but it's equal to y, which is specified static, and with how we currently do things we won't promote y to dynamic, but will demote(?) x to static. So this can be surprising if you don't fully know your model, and/or missed one of your other inputs when specifying auto-dynamic shapes.
```
class Foo(torch.nn.Module):
    def forward(self, x, y):
        return x + y
inputs = (torch.randn(6), torch.randn(6))
export(Foo(), inputs, dynamic_shapes={"x": (DIM.AUTO,), "y": None})
```

Caveat 2: specifying `AUTO` and Dims in the same spec is still problematic:

- The way Dims/DerivedDims are currently handled is very strict. A Dim represents a symbol, and we require a user to specify the symbol for all dims governed by the symbol - that's why we've seen errors in the past like `The values of x must always be related to y by ...`, asking the user to specify the exact relation as in the program. We also require the specified min/max range to be a subset of the valid range from model analysis. All this doesn't compose well with specifying `AUTO` just yet - for example in the program below, ideal behavior could be to return a dynamic program, where `dx = x.size(0) = y.size(0)` has range (3,6). Unfortunately this crashes, and correct behavior is to specify `dx` for both inputs. So currently we raise a UserError and crash if both Dims + `AUTO` are present in the spec.
```
class Foo(torch.nn.Module):
    def forward(self, x, y):
        return x + y
inputs = (torch.randn(6), torch.randn(6))
export(Foo(), inputs, dynamic_shapes={"x": (DIM.AUTO,), "y": {0: Dim("dx", min=3, max=6)}})  # this doesn't work, because x & y and related
```

Implementation details:

This is done by setting `assume_static_by_default=False`, and doing a transform on the `dynamic_shapes` spec to preserve semantics. `assume_static_by_default=False` will treat unspecified dims or Nones as dynamic. This is the opposite of what `export.export()` currently does - unspecified Dims/Nones are treated as static. Historically this static-by-default behavior, where the user deals with fewer guards, has been desirable, and we would like to respect that in this implementation. So this internal spec transformation is added, `_transform_shapes_for_default_dynamic()`, does the spec conversion necessary to be compatbile with dynamic by default. Specifically, AUTOs are converted into Nones, and Nones/unspecified dims are filled in with explicitly static constraints. 

For example, this would look like, for a 3-d tensor: `{0: DIM.AUTO, 1: None, 2: Dim("dx")} -> {0: None, 1: 32, 2: Dim("dx")}`

This does seem overly complicated, but it's done to preserve dynamic shapes semantics for `torch._dynamo.export()`, which already uses `assume_static_by_default=False`, and follows the same process for generating shape constraints , via `_process_dynamic_shapes`. There the semantics are:
```
None/unspecified: dynamic by default
Dim/DerivedDim: also a strict assertion
```

If we don't care about BC for `_dynamo.export(dynamic_shapes)`, then we can just modify semantics for `_process_dynamic_shapes()` and change all the relevant tests in `test/dynamo/test_export.py`.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames


Reviewed By: avikchaudhuri

Differential Revision: D61352432

Pulled By: pianpwk
